File: app/src/main/java/com/imangazalievm/bubbble/presentation/mvp/presenters/UserDetailsPresenter.java
package com.imangazalievm.bubbble.presentation.mvp.presenters;

import com.arellomobile.mvp.InjectViewState;
import com.arellomobile.mvp.MvpPresenter;
import com.imangazalievm.bubbble.domain.exceptions.NoNetworkException;
import com.imangazalievm.bubbble.domain.interactors.UserDetailsInteractor;
import com.imangazalievm.bubbble.domain.models.User;
import com.imangazalievm.bubbble.presentation.commons.rx.RxSchedulersProvider;
import com.imangazalievm.bubbble.presentation.mvp.views.UserDetailsView;
import com.imangazalievm.bubbble.presentation.utils.DebugUtils;

import javax.inject.Inject;

@InjectViewState
public class UserDetailsPresenter extends MvpPresenter<UserDetailsView> {

    private UserDetailsInteractor userDetailsInteractor;
    private RxSchedulersProvider rxSchedulersProvider;
    private long userId;
    private User user;

    @Inject
    public UserDetailsPresenter(UserDetailsInteractor userDetailsInteractor,
                                RxSchedulersProvider rxSchedulersProvider,
                                long userId) {
        this.userDetailsInteractor = userDetailsInteractor;
        this.rxSchedulersProvider = rxSchedulersProvider;
        this.userId = userId;

        loadUser();
    }

    private void loadUser() {
        userDetailsInteractor.getUser(userId)
                .compose(rxSchedulersProvider.getIoToMainTransformerSingle())
                .subscribe(this::onUserLoadSuccess, this::onUserLoadError);
    }

    private void onUserLoadSuccess(User user) {
        this.user = user;
        getViewState().hideLoadingProgress();
        getViewState().showUser(user);
    }

    private void onUserLoadError(Throwable throwable) {
        if (throwable instanceof NoNetworkException) {
            getViewState().hideLoadingProgress();
            getViewState().showNoNetworkLayout();
        } else {
            DebugUtils.showDebugErrorMessage(throwable);
        }
    }

    public void retryLoading() {
        getViewState().hideNoNetworkLayout();
        getViewState().showLoadingProgress();
        loadUser();
    }

    public void onUserTwitterButtonClicked() {
        getViewState().openInBrowser(user.getLinks().getTwitter());
    }

    public void onUserWebsiteButtonClicked() {
        getViewState().openInBrowser(user.getLinks().getWeb());
    }

}

File: app/src/main/java/com/bubbble/presentation/userprofile/details/UserDetailsPresenter.kt
package com.bubbble.presentation.userprofile.details

import com.arellomobile.mvp.InjectViewState
import com.bubbble.data.global.network.exceptions.NoNetworkException
import com.bubbble.domain.global.models.User
import com.bubbble.domain.userprofile.UserDetailsInteractor
import com.bubbble.presentation.global.mvp.BasePresenter
import dagger.assisted.Assisted
import dagger.assisted.AssistedFactory
import dagger.assisted.AssistedInject

@InjectViewState
class UserDetailsPresenter @AssistedInject constructor(
    private val userDetailsInteractor: UserDetailsInteractor,
    @Assisted  private val userId: Long
) : BasePresenter<UserDetailsView>() {

    private lateinit var user: User
    private val isUserLoaded: Boolean
        get() = ::user.isInitialized

    override fun onFirstViewAttach() {
        super.onFirstViewAttach()
        loadUser()
    }

    private fun loadUser() = launchSafe {
        viewState.showLoadingProgress()
        try {
            this@UserDetailsPresenter.user = userDetailsInteractor.getUser(userId)
            viewState.showUserInfo(user)
        } catch (e: NoNetworkException) {
            viewState.showNoNetworkLayout()
        } finally {
            viewState.hideLoadingProgress()
        }
    }

    fun retryLoading() {
        viewState.hideNoNetworkLayout()
        loadUser()
    }

    fun onUserTwitterButtonClicked() {
        if (!isUserLoaded) return
        viewState.openInBrowser(user.links.twitter)
    }

    fun onUserWebsiteButtonClicked() {
        if (!isUserLoaded) return
        viewState.openInBrowser(user.links.web)
    }

    @AssistedFactory
    interface Factory {
        fun create(userId: Long): UserDetailsPresenter
    }

}